### DISCUSSION TITLE
Rclip used for grouping pictures by context

### DISCUSSION THREAD

**Discussion by LamprosPitsillos** [yurijmikhalevich/rclip#100] [General]
Google photos has a feature where you can group pictures by people/faces.
I was wondering if `rclip` is able to do something or any similar tool can be used

**Comment by yurijmikhalevich** [comment]
Hi @LamprosPitsillos. Currently, rclip doesn't support this, but I have tentatively put this feature on rclip's roadmap here: https://github.com/yurijmikhalevich/rclip/issues/101.

**Reply by LamprosPitsillos** [reply]
Thank you so much!!
A few comments:
Since going through the trouble, i would suggest a broader feature first of grouping by all sorts of tags ( mountain, food , family etc)
I was wandering if the indexing that is being done by `rclip` exposes or could expose things like that ?
So then the user could just query the index (or a more detailed db)

**Reply by yurijmikhalevich** [reply]
@LamprosPitsillos, what do you mean by "grouping"?

**Reply by LamprosPitsillos** [reply]
![Screenshot_20240414_135306_Photos](https://github.com/yurijmikhalevich/rclip/assets/61395246/e68a5c24-22c3-4153-a2d6-c598001cd58e)

Google pictures for example has a search feature where it groups things according to the subject

**Reply by yurijmikhalevich** [reply]
@LamprosPitsillos, since `rclip` is a terminal tool, how do you imagine the interaction with `rclip` doing this?

Currently, you can do `rclip <query>` and get a "group" of images matching the query, which is more flexible than having a set of predefined groups.

There is a sister project, [`rtag`](https://github.com/yurijmikhalevich/rtag) that will enable users to write "tags" into the image metadata for other software to use. Maybe this can solve what you are trying to do?

**Reply by LamprosPitsillos** [reply]
Thanks for the recommendation,the project seems close to what i want.
> how do you imagine the interaction with rclip doing this?

Not graphically for sure.
I imagined 2 ways:
- the most maintainable and more user orientated approach is to tag the images like i imagine the indexing does and have the user just query the index ex. 
```bash
sqlite3 ./index ' select path where tag == "snow" ' |  *image-viewer* -
```
- the second thought is to run `rclip` without arguments and it shows all the available tags in the directory having the user choose one and it returns the paths like usual. This will require more changes and is more of UX sigar for the first one that only works per directory.

This are just a few ways i imagined this functionality.

Edit: 
I now see that this example was very confusing as i was just describing the standard use of `rclip` with extra steps...

Please see my later comment and below i will try to provide a better example.

```bash
sqlite3 ./index ' select district tag from tags' |  fzf | xargs rclip ...
```

**Reply by yurijmikhalevich** [reply]
> sqlite3 ./index ' select path where tag == "snow" ' |  *image-viewer* -

@LamprosPitsillos, why not just run?

```
rclip -t 100 --no-indexing --preview snow
```

"no-indexing" will make rclip skip checking files for changes, which is the slowest part when querying things.

Not that `--preview` works only in [iTerm2](https://iterm2.com/), [Konsole](https://konsole.kde.org/) (version 22.04 and higher), [wezterm](https://wezfurlong.org/wezterm/), [Mintty](https://mintty.github.io/), or [mlterm](https://mlterm.sourceforge.net/).

**Reply by yurijmikhalevich** [reply]
Or, it's also possible to:
```
rclip -t 100 -fn snow | *image-viewer* -
```

`-f` makes it output the filepaths only, so it's easier to pipe the output to the image-viewer
`-n` is short for `--no-indexing`
`-t 100` makes it output top 100 results, you can use any number here

**Reply by LamprosPitsillos** [reply]
My current scenario is that i have a Huge collection of images, that are un categorised.
I don't have a query to search for like "snow" or "animals" . I have forgotten or the time frame of all the images is to large to search for something specific.

The issue is not the speed at which it indexes as you have mentioned. The issue is that i am trying to eliminate the "query" from the user .

Imagine a scenario where in a large image gallery, somewhere hidden are pictures of my vacation to the Himalayan mountains that i have *completely* forgotten about.
As i do not have any recollection of my vacation to the mountains i will never provide s query like "mountains" "snow" "skiing" etc 

Or a more simple scenario, also one of my intended use cases. I have found a drive with thousands of old pictures. I have no idea of what to queries them for.

I imagine , all of this fit better in the scope of a more mature `rtag` down the line, but since rclip is more for searching images i think it is also appropriate for this project as well

**Reply by yurijmikhalevich** [reply]
> I have found a drive with thousands of old pictures. I have no idea of what to queries them for.

So what I am hearing is that you are looking for a tool that will let you explore an unknown set of images/describe a set of images for you. Is it right?

**Reply by LamprosPitsillos** [reply]
Yes , you got that right 👍🏻
Explore + help with sorting them in groups and what not with the given data

**Reply by yurijmikhalevich** [reply]
@LamprosPitsillos, I see. I will think about it. What do you think of if you could run a command like `rclip explore` and it would spit out top keywords associated with the images in the current directory + let you pick any, and then give you a list of images matching the keyword?

**Reply by LamprosPitsillos** [reply]
That sounds great! Thank you for considering it !
A few notes:
- i would like for a way to view all keywords to avoid small groups of pictures being buried ( a few pictures of a "party" from a galery of thousands will get lost in a sea of other tags)
- reserved

**Reply by yurijmikhalevich** [reply]
@LamprosPitsillos, thank you for your thoughts and feedback :) What do you mean by "reserved"?

**Reply by LamprosPitsillos** [reply]
Nothing nothing, i left it as reserved cause i thought i had more to add, but nothing else came up , sorry 😅
